kork/kork-artifacts/src/main/java/com/netflix/spinnaker/kork/artifacts/artifactstore/entities/ManifestStorageCollectionHandler.java

```diff
@@ -0,0 +1,106 @@
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.kork.artifacts.artifactstore.entities;
+
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.kork.artifacts.ArtifactTypeDecorator;
+import com.netflix.spinnaker.kork.artifacts.ArtifactTypes;
+import com.netflix.spinnaker.kork.artifacts.artifactstore.ArtifactStore;
+import com.netflix.spinnaker.kork.artifacts.artifactstore.filters.ApplicationFilter;
+import com.netflix.spinnaker.kork.artifacts.artifactstore.filters.ApplicationStorageFilter;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import com.netflix.spinnaker.security.AuthenticatedRequest;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * ManifestStorageHandler is used to convert manifest values into artifacts. This is done by
+ * matching against a key within the execution context. Upon matching, we will convert the manifest
+ * into an artifact and write the artifact in its place.
+ */
+public class ManifestStorageCollectionHandler implements ArtifactStoragePropertyHandler {
+  private List<String> keys = List.of("manifests");
+  private final ApplicationFilter exclude;
+
+  public ManifestStorageCollectionHandler(Map<String, List<ApplicationStorageFilter>> exclude) {
+    this.exclude =
+        new ApplicationFilter(exclude.get(ArtifactTypes.EMBEDDED_MAP_BASE64.getMimeType()));
+  }
+
+  private <K, V> Map<K, V> handleMap(ArtifactStore store, Map<K, V> v) {
+    return this.convert(store, v);
+  }
+
+  private <T> T convert(ArtifactStore store, T t) {
+    Map<?, ?> v = (Map<?, ?>) t;
+    if (EntityHelper.alreadyStored(v)) {
+      return t;
+    }
+
+    Artifact artifact = EntityHelper.toArtifact(v, ArtifactTypes.EMBEDDED_MAP_BASE64.getMimeType());
+    Artifact stored = store.store(artifact, ArtifactTypeDecorator.toRemote(artifact));
+    if (ArtifactTypes.EMBEDDED_MAP_BASE64.getMimeType().equals(stored.getType())) {
+      return t;
+    }
+    // We can cast directly to T due to our check of canHandle.
+    return (T) EntityHelper.toMap(stored);
+  }
+
+  @Override
+  public boolean canHandleProperty(BeanProperty property, Object value) {
+    if (AuthenticatedRequest.getSpinnakerExecutionId().isEmpty() || this.exclude.shouldFilter()) {
+      return false;
+    }
+
+    String propertyName = property.getName();
+    Collection<?> v = (Collection<?>) value;
+    if (!(v instanceof Set && v.size() != 0)) {
+      return false;
+    }
+
+    Object obj = v.iterator().next();
+    if (!(obj instanceof Map)) {
+      return false;
+    }
+    // Good ol' clouddriver sometimes stores manifests as a list and other times as a set.
+    return this.keys.stream().anyMatch(k -> k.equals(propertyName))
+        && !EntityHelper.alreadyStored((Map) obj);
```

Comment:
I'm surprised to see something dynamic like `alreadyStored` in a method like canHandleProperty.  Couldn't this code handle the object whether it's stored or not?